### PR TITLE
Setup reverse proxy for remote users

### LINKED ISSUE #9: Compatibility issues with google collab

### LINKED ISSUE #82: ConfyUI using domain
I have my Stable Diffusion "public" and was trying to have ConfyUI on it, but I can't configure the page the tab is trying to go, so it goes to the local host. is there any way I can redirect it to my domain or do I have to do a workaround?

In the machine where I am running Stable Diffusion it all works as intended.

image

### LINKED ISSUE #107: sd-webui-comfy not loading in webui tab remotely.
Hey, I am unsure if this is similar to the domain issue from about 2 weeks back but I am having an issue when attempting to load comfyui through the webui remotely. This is within the vlad/SD.Next, windows 11. All installed fine no error outputs in the terminal and it accepts the 0.0.0.0:8189 input for the cli webui argument to point at the ComfyUi.

Screenshot of what occurs:
ComfyUI wrbui Tab

Terminal output:
PS D:\AI Stuff\Vlad1111> .\webui.bat --listen --comfyui-listen
Using VENV: D:\AI Stuff\Vlad1111\venv
10:25:03-696502 INFO     Starting SD.Next
10:25:03-702502 INFO     Python 3.10.6 on Windows
10:25:03-726502 INFO     Version: 9d17cf4c Mon Aug 7 19:50:39 2023 +0300
10:25:04-188180 INFO     Latest published version: ff2d846667f62bd83a49b85e44e137e97c64727b 2023-08-07T21:51:23Z
10:25:04-191699 INFO     nVidia CUDA toolkit detected
10:25:04-269290 WARNING  Modified files: ['modules/lora', 'modules/lycoris']
10:25:04-294546 INFO     Disabled extensions: ['sd-webui-agent-scheduler']
10:25:04-295547 INFO     Enabled extensions-builtin: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding', 'sd-extension-system-info',
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
10:25:04-297546 INFO     Enabled extensions: ['a1111-sd-webui-tagcomplete', 'novelai-2-local-prompt', 'OneButtonPrompt', 'openpose-editor',
                         'sd-canvas-editor', 'sd-webui-3d-open-pose-editor', 'sd-webui-comfyui', 'sd-webui-infinite-image-browsing',
                         'sd-webui-inpaint-anything', 'sd-webui-regional-prompter', 'stable-diffusion-webui-aesthetic-image-scorer',
                         'Stable-Diffusion-Webui-Civitai-Helper']
10:25:04-301064 INFO     Verifying requirements
10:25:04-308064 INFO     Verifying packages
10:25:04-313571 INFO     Verifying repositories
10:25:08-099765 INFO     Verifying submodules
10:25:23-322068 INFO     Extensions enabled: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding', 'sd-extension-system-info',
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR',
                         'a1111-sd-webui-tagcomplete', 'novelai-2-local-prompt', 'OneButtonPrompt', 'openpose-editor', 'sd-canvas-editor',
                         'sd-webui-3d-open-pose-editor', 'sd-webui-comfyui', 'sd-webui-infinite-image-browsing', 'sd-webui-inpaint-anything',
                         'sd-webui-regional-prompter', 'stable-diffusion-webui-aesthetic-image-scorer',
                         'Stable-Diffusion-Webui-Civitai-Helper']
10:25:23-323068 INFO     Verifying packages
10:25:23-327068 INFO     Extension preload: 0.0s D:\AI Stuff\Vlad1111\extensions-builtin
10:25:23-328068 INFO     Extension preload: 0.0s D:\AI Stuff\Vlad1111\extensions
10:25:23-335521 INFO     Server arguments: ['--listen', '--comfyui-listen']
No module 'xformers'. Proceeding without it.
10:25:27-056966 INFO     Pipeline: Backend.ORIGINAL
10:25:27-419001 INFO     Libraries loaded
10:25:27-420002 INFO     Using data path: D:\AI Stuff\Vlad1111
10:25:27-423508 INFO     Available VAEs: D:\AI Stuff\Model Storage\models\VAE 5
10:25:27-462423 INFO     Available models: D:\AI Stuff\Model Storage\models\Stable-diffusion 98
10:25:28-522759 INFO     ControlNet v1.1.233
ControlNet v1.1.233
ControlNet preprocessor location: D:\AI Stuff\Vlad1111\extensions-builtin\sd-webui-controlnet\annotator\downloads
10:25:28-601745 INFO     ControlNet v1.1.233
ControlNet v1.1.233
Tag Autocomplete: Could not locate model-keyword extension, LORA/LYCO trigger word completion will be unavailable.
[sd-webui-comfyui] restoring graceful SIGINT handler for the webui process
Civitai Helper: Get Custom Model Folder
Civitai Helper: Load setting from: D:\AI Stuff\Vlad1111\extensions\Stable-Diffusion-Webui-Civitai-Helper\setting.json
Civitai Helper: No setting file, use default
10:25:34-064803 INFO     Loading UI theme: name=gradio/default style=Dark
10:25:34-103894 INFO     Extra network created thumbnails: checkpoints 0
Running on local URL:  http://0.0.0.0:7860
10:25:40-360323 INFO     Local URL: http://localhost:7860/
10:25:40-361323 INFO     Initializing middleware
[sd-webui-comfyui] started callback listeners for process webui
Loading weights: D:\AI Stuff\Model Storage\models\Stable-diffusion\bb95FurryMix_v110.safetensors ━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/4.3 GB -:--:--10:25:41-246073 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16 unet=torch.float16
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
[ComfyUI] [sd-webui-comfyui] started callback listeners for process comfyui
[ComfyUI] [sd-webui-comfyui] Injecting custom extensions...
[ComfyUI] Total VRAM 24564 MB, total RAM 65277 MB
Loading weights: D:\AI Stuff\Model Storage\models\VAE\vae-ft-mse-840000-ema-pruned.safetensors ━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0/334.6 MB -:--:--10:25:42-759007 INFO     Applying scaled dot product cross attention optimization
10:25:42-790859 INFO     Embeddings: loaded=28 skipped=5
10:25:42-795859 INFO     Model loaded in 1.8s (load=0.3s create=0.2s apply=0.5s vae=0.5s move=0.4s)
10:25:42-984319 INFO     Model load finished: {'ram': {'used': 5.32, 'total': 63.75}, 'gpu': {'used': 4.19, 'total': 23.99}, 'retries': 0,
                         'oom': 0} cached=0
10:25:43-449493 INFO     Startup time: 20.1s (torch=2.6s gradio=0.5s libraries=1.0s scripts=6.3s onchange=0.3s ui-txt2img=0.3s ui-img2img=0.1s
                         ui-settings=0.2s ui-extensions=1.4s launch=4.3s app-started=0.2s checkpoint=2.8s)
[ComfyUI] Set vram state to: NORMAL_VRAM
[ComfyUI] Device: cuda:0 NVIDIA GeForce RTX 3090 Ti : native
[ComfyUI] Using pytorch cross attention
[ComfyUI] [sd-webui-comfyui] Launching ComfyUI with arguments: --listen 0.0.0.0 --port 8189
[ComfyUI]
Import times for custom nodes:
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_save_image.py
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_io.py
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_checkpoint_loader.py
[ComfyUI]
[ComfyUI] Starting server

11:50:16-276332 INFO     Exiting
[sd-webui-comfyui] stopped callback listeners for process webui
Terminate batch job (Y/N)? y
PS D:\AI Stuff\Vlad1111> .\webui.bat --listen --comfyui-listen --webui-comfyui-client-address 0.0.0.0:8189
Using VENV: D:\AI Stuff\Vlad1111\venv
11:50:55-809512 INFO     Starting SD.Next
11:50:55-815061 INFO     Python 3.10.6 on Windows
11:50:55-844743 INFO     Version: 9d17cf4c Mon Aug 7 19:50:39 2023 +0300
11:50:56-548215 INFO     Latest published version: ff2d846667f62bd83a49b85e44e137e97c64727b
                         2023-08-07T21:51:23Z
11:50:56-552263 INFO     nVidia CUDA toolkit detected
11:50:56-614961 WARNING  Modified files: ['modules/lora', 'modules/lycoris']
11:50:56-640775 INFO     Disabled extensions: ['sd-webui-agent-scheduler']
11:50:56-641777 INFO     Enabled extensions-builtin: ['a1111-sd-webui-lycoris',
                         'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET',
                         'sd-dynamic-thresholding', 'sd-extension-system-info',
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser',
                         'stable-diffusion-webui-rembg', 'SwinIR']
11:50:56-643861 INFO     Enabled extensions: ['a1111-sd-webui-tagcomplete',
                         'novelai-2-local-prompt', 'OneButtonPrompt', 'openpose-editor',
                         'sd-canvas-editor', 'sd-webui-3d-open-pose-editor',
                         'sd-webui-comfyui', 'sd-webui-infinite-image-browsing',
                         'sd-webui-inpaint-anything', 'sd-webui-regional-prompter',
                         'stable-diffusion-webui-aesthetic-image-scorer',
                         'Stable-Diffusion-Webui-Civitai-Helper']
11:50:56-647312 INFO     Verifying requirements
11:50:56-656827 INFO     Verifying packages
11:50:56-662827 INFO     Verifying repositories
11:51:00-550719 INFO     Verifying submodules
11:51:20-808695 INFO     Extensions enabled: ['a1111-sd-webui-lycoris',
                         'clip-interrogator-ext', 'LDSR', 'Lora',
                         'multidiffusion-upscaler-for-automatic1111', 'ScuNET',
                         'sd-dynamic-thresholding', 'sd-extension-system-info',
                         'sd-webui-controlnet', 'stable-diffusion-webui-images-browser',
                         'stable-diffusion-webui-rembg', 'SwinIR',
                         'a1111-sd-webui-tagcomplete', 'novelai-2-local-prompt',
                         'OneButtonPrompt', 'openpose-editor', 'sd-canvas-editor',
                         'sd-webui-3d-open-pose-editor', 'sd-webui-comfyui',
                         'sd-webui-infinite-image-browsing', 'sd-webui-inpaint-anything',
                         'sd-webui-regional-prompter',
                         'stable-diffusion-webui-aesthetic-image-scorer',
                         'Stable-Diffusion-Webui-Civitai-Helper']
11:51:20-810921 INFO     Verifying packages
11:51:20-814007 INFO     Extension preload: 0.0s D:\AI Stuff\Vlad1111\extensions-builtin
11:51:20-815007 INFO     Extension preload: 0.0s D:\AI Stuff\Vlad1111\extensions
11:51:20-839614 INFO     Server arguments: ['--listen', '--comfyui-listen',
                         '--webui-comfyui-client-address', '0.0.0.0:8189']
No module 'xformers'. Proceeding without it.
11:51:25-556295 INFO     Pipeline: Backend.ORIGINAL
11:51:26-390568 INFO     Libraries loaded
11:51:26-391568 INFO     Using data path: D:\AI Stuff\Vlad1111
11:51:26-395568 INFO     Available VAEs: D:\AI Stuff\Model Storage\models\VAE 5
11:51:27-169707 INFO     Available models: D:\AI Stuff\Model
                         Storage\models\Stable-diffusion 98
11:51:28-303101 INFO     ControlNet v1.1.233
ControlNet v1.1.233
ControlNet preprocessor location: D:\AI Stuff\Vlad1111\extensions-builtin\sd-webui-controlnet\annotator\downloads
11:51:28-512422 INFO     ControlNet v1.1.233
ControlNet v1.1.233
Tag Autocomplete: Could not locate model-keyword extension, LORA/LYCO trigger word completion will be unavailable.
[sd-webui-comfyui] restoring graceful SIGINT handler for the webui process
Civitai Helper: Get Custom Model Folder
Civitai Helper: Load setting from: D:\AI Stuff\Vlad1111\extensions\Stable-Diffusion-Webui-Civitai-Helper\setting.json
Civitai Helper: No setting file, use default
11:51:34-704611 INFO     Loading UI theme: name=gradio/default style=Dark
11:51:35-031390 INFO     Extra network created thumbnails: checkpoints 0
Running on local URL:  http://0.0.0.0:7860
11:51:41-982860 INFO     Local URL: http://localhost:7860/
11:51:41-984867 INFO     Initializing middleware
[sd-webui-comfyui] started callback listeners for process webui
Loading weights: D:\AI Stuff\Model Storage\models\Stable-diffusion\bb95FurryMix_v110.safe…
11:51:42-875012 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16
                         unet=torch.float16
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
[ComfyUI] [sd-webui-comfyui] started callback listeners for process comfyui
[ComfyUI] [sd-webui-comfyui] Injecting custom extensions...
[ComfyUI] Total VRAM 24564 MB, total RAM 65277 MB
Loading weights: D:\AI Stuff\Model Storage\models\VAE\vae-ft-mse-840000-ema-pruned.safete…
11:51:44-449731 INFO     Applying scaled dot product cross attention optimization
11:51:44-982107 INFO     Embeddings: loaded=28 skipped=5
11:51:44-986107 INFO     Model loaded in 2.4s (load=0.3s create=0.2s apply=0.5s vae=0.5s
                         move=0.4s embeddings=0.5s)
11:51:45-176018 INFO     Model load finished: {'ram': {'used': 5.31, 'total': 63.75},
                         'gpu': {'used': 4.19, 'total': 23.99}, 'retries': 0, 'oom': 0}
                         cached=0
[ComfyUI] Set vram state to: NORMAL_VRAM
[ComfyUI] Device: cuda:0 NVIDIA GeForce RTX 3090 Ti : native
[ComfyUI] Using pytorch cross attention
[ComfyUI] [sd-webui-comfyui] Launching ComfyUI with arguments: --listen 0.0.0.0 --port 8189
11:51:45-658939 INFO     Startup time: 24.8s (torch=2.6s gradio=1.0s libraries=1.9s
                         models=0.8s scripts=7.2s onchange=0.3s ui-txt2img=1.1s
                         ui-img2img=0.1s ui-settings=0.2s ui-extensions=1.6s
                         ui-defaults=0.1s launch=4.3s app-started=0.2s checkpoint=3.4s)
[ComfyUI]
Import times for custom nodes:
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_save_image.py
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_io.py
[ComfyUI]    0.0 seconds: D:\AI Stuff\Vlad1111\extensions\sd-webui-comfyui\comfyui_custom_nodes\webui_checkpoint_loader.py
[ComfyUI]
[ComfyUI] Starting server

[ComfyUI] To see the GUI go to: http://0.0.0.0:8189
=========================

No issues with ComfyUI itself when held by SD.Next/Vlad runs perfectly, just the output in the tab does not display while using the --listen hook to push both SD.Next and ComfyUI through VPN. I have tried giving a straight IP for what the pc its hosted on is but no dice. Any help appreciated :).

### PR DESCRIPTION
**Pull request by ljleb**
closes #107, closes #82, closes #9

Needs `--api` to create the reverse proxy.

### CODE REVIEW

**Comment by ljleb** [comment]
Maybe should detect whether the webui is launched with a reverse proxy to automatically toggle the comfyui reverse proxy

File: lib_comfyui/webui/reverse_proxy.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                (proxy_route_bytes * 2, proxy_route_bytes),
+            ]
+            for substring, replacement in replacements:
+                chunk = chunk.replace(substring, replacement)
+            yield chunk
+
+    web_client = httpx.AsyncClient(base_url=comfyui_url)
+
+    async def reverse_proxy(request: Request):
+        """Proxy incoming requests to another server."""
```
Comment (ljleb) [inline lib_comfyui/webui/reverse_proxy.py:39]: remove the docstrings or fix them

**Comment by John-WL** [comment]
does this implementation require the `--api` arg?

**Comment by John-WL** [comment]
Title should be ##, not just #.

**Review by John-WL — APPROVED** [review]

**Pull request merged**